scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkBackend.scala

```diff
@@ -130,19 +132,157 @@ object SparkPlanner {
           val op = rec(pipe)
-          ???
-        case (ReduceStepPipe(IteratorMappedReduce(_, pipe, fn, _, _)), rec) =>
+          op.sortedMapGroup(fn)(ordK, ordV)
+        case (ReduceStepPipe(IteratorMappedReduce(ordK, pipe, fn, _, _)), rec) =>
           val op = rec(pipe)
-          ???
+          op.mapGroup(fn)(ordK)
       }
     })
+
+  case class OnEmptyIterator[A](it: Iterator[A], fn: () => Unit) extends Iterator[A] {
+    var fnMut: () => Unit = fn
+    def hasNext = it.hasNext || {
+      if (fnMut != null) { fnMut(); fnMut = null }
+      false
+    }
+
+    def next = it.next
+  }
+
+  case class CachingSum[K, V](capacity: Int, semigroup: Semigroup[V]) extends Function1[Iterator[(K, V)], Iterator[(K, V)]] {
+    def newCache(evicted: MMap[K, V]): JMap[K, V] = new JLinkedHashMap[K, V](capacity + 1, 0.75f, true) {
+      override protected def removeEldestEntry(eldest: JMap.Entry[K, V]) =
+        if (super.size > capacity) {
+          evicted.put(eldest.getKey, eldest.getValue)
+          true
+        } else {
+          false
+        }
+    }
+
+    def apply(kvs: Iterator[(K, V)]) = {
+      val evicted = MMap.empty[K, V]
+      val currentCache = newCache(evicted)
+      // TODO actually do something here
+      new Iterator[(K, V)] {
+        var resultIterator: Iterator[(K, V)] = Iterator.empty
+
+        def hasNext = kvs.hasNext || resultIterator.hasNext
+        @annotation.tailrec
+        def next: (K, V) = {
+          if (resultIterator.hasNext) {
+            resultIterator.next
+          } else if (kvs.hasNext) {
+            val (k, deltav) = kvs.next
+            val vold = currentCache.get(k)
+            if (vold == null) {
+              currentCache.put(k, deltav)
+            } else {
+              currentCache.put(k, semigroup.plus(vold, deltav))
+            }
+            // let's see if we have anything to evict
+            if (evicted.nonEmpty) {
+              resultIterator = OnEmptyIterator(evicted.iterator, () => evicted.clear())
+            }
+            next
+          } else {
+            // time to flush the cache
+            import scala.collection.JavaConverters._
+            val cacheIter = currentCache.entrySet.iterator.asScala.map { e => (e.getKey, e.getValue) }
+            resultIterator = OnEmptyIterator(cacheIter, () => currentCache.clear())
+            next
+          }
+        }
+      }
+    }
+  }
+
+  private def planHashJoinable[K, V](hj: HashJoinable[K, V], rec: FunctionK[TypedPipe, Op]): Op[(K, V)] =
+    rec(TypedPipe.ReduceStepPipe(HashJoinable.toReduceStep(hj)))
+
+  private def planCoGroup[K, V](config: Config, cg: CoGrouped[K, V], rec: FunctionK[TypedPipe, Op]): Op[(K, V)] = {
+    import CoGrouped._
+
+    cg match {
+      case FilterKeys(cg, fn) =>
+        planCoGroup(config, cg, rec).filter { case (k, _) => fn(k) }
+      case MapGroup(cg, fn) =>
+        // don't need to repartition, just mapPartitions
```

Comment:
Do we not need to call the MapGroup operator here? or are we guaranteed that its already been cogrouped/shuffled?